src/snowflake/snowpark/mock/_stage_registry.py

```diff
@@ -413,6 +449,96 @@ def read_file(
+                        content = json.load(file, cls=CUSTOM_JSON_DECODER)
+                        tmp_content = {}
+                        # snowflake escape double quotes by adding extra double quote
+                        for key, value in content.items():
+                            key = key.replace('"', '""')
+                            tmp_content[f'"{key}"'] = value
+                        content = tmp_content
+                        contents.append(content)
+                        # extract the schema from the content
+                        for column_name, value in content.items():
+                            # snowflake double quote column name read from json file
+                            target_datatype = infer_type(value)
+                            # multiple json files can be of different schema
+                            # if we find an existing schema but type is different from the inferred one
+                            # we convert the column datatype to string, this is snowflake behavior
+                            if column_name in result_df_sf_types and not isinstance(
+                                target_datatype,
+                                type(result_df_sf_types[column_name].datatype),
+                            ):
+                                # snowflake convert to string type
+                                target_datatype = StringType()
+
+                            column_series = ColumnEmulator(
+                                data=None,
+                                dtype=object,
+                                name=column_name,
+                                sf_type=ColumnType(target_datatype, nullable=True),
+                            )
+                            result_df[column_name], result_df_sf_types[column_name] = (
+                                column_series,
+                                column_series.sf_type,
+                            )
+                # fill empty cells with None value, this aligns with snowflake
+                for content in contents:
+                    for miss_key in set(result_df_sf_types.keys()) - set(
+                        content.keys()
+                    ):
+                        content[miss_key] = None
+                    df = TableEmulator([content])
+                    result_df = pd.concat([result_df, df], ignore_index=True)
+                # when concat is called, sf_type information gets lost, so we reset the type info in the end
+                result_df.sf_types = result_df_sf_types
+
+                # in the case that there are values of different types in the same column, snowflake will
+                # convert data into string
+                for col_name in result_df.columns:
+                    if isinstance(result_df_sf_types[col_name].datatype, StringType):
+                        result_df[col_name] = mock_to_char(result_df[col_name])
```

Comment:
Since we the column is of mixed types now, lets's set `result_df_sf_types[col_name].datatype` to `VariantType` then use `mock_to_char`.

Reply:
good idea, code updated